Feature: OPA Entity Checker Catalog Processor Module - validate all entities always (#248)
Parsifal-M/backstage-opa-plugins: docs/opa-entity-checker/introduction.md

![Compact MetaData Card Violations Open](../assets/card-compact-opened.png)

## Entity Processor Validation - Validate all entities always

> While the frontend validation allows nice feedback for users, only entities that are visited will be validated. After implementing new rules, you might wonder how many entities are now returning validation error.

Reviewer: ```suggestion
While the frontend validation allows for quick feedback, users will only see the feedback on visiting an entity. After implementing new rules, you might wonder how many entities are now returning validation error.
```

> This is where the catalog processor comes into play. When Backstage discover entities in your repositories it allows plugins called `porcessor` to validate and modify entities. The OPA Entity Checker Catalog Processor will validate your entities whenever they are ingested and add annotation reflecting it's result.

Reviewer: ```suggestion
This is where catalog processors comes into play. When you discover or import entities into the Backstage catalog, you can run `processors` to (in this case) validate and update entities. The OPA Entity Checker Catalog Processor will validate your entities when they are ingested into the catalog and add an annotation based on the evaluation of the policy!
```

For instance, the final entity in Backstage might look like this:

```yaml
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  namespace: default
  annotations:
    open-policy-agent/entity-checker-violations-error-count: '2'
    open-policy-agent/entity-checker-violations-warning-count: '1'
    open-policy-agent/entity-checker-violations-status: error
```

> This means that your validation reported 2 errors and 1 warning. The `status` of the validation is thus `error`. This can be very handy because now we have a way to query the catalogue to find those entities that violate validation.

Reviewer: ```suggestion
As you can see, an annotation of `open-policy-agent/entity-checker-violations-status: error` has been added to the entity, this is because after the processor ran and checked the metadata against the OPA policy it found a violation which is considered an `error`
```
